### ISSUE TITLE
Unpacking after reshaping can cause inaccurate packed sampling

### ISSUE THREAD

**Issue by annxingyuan** [tensorflow/tfjs#834]
cc @caisq @nsthorat @dsmilkov 

This bug was introduced by https://github.com/tensorflow/tfjs-core/pull/1329

We impose a constraint on packed texels that they may contain only adjacent rows / columns from the same batch. This means if we reshape a tensor backed by a packed texture, we may also need to rearrange the values within the texture so that the constraint is maintained. For example, if we reshape a tensor from `[9, 1]` to `[3, 3, 1]`, the texture grows from 5 texels to 6 texels:

`[9, 1]` vs `[3, 3, 1]`:

```
0|x          0|x
---          ---
1|x          1|x

2|x          2|x
---          ---
3|x          x|x

4|x          3|x
---          ---
5|x          4|x

6|x          5|x
---          ---
7|x          x|x

8|x          6|x
---          ---
x|x          7|x
               
             8|x
             ---
             x|x
```

Otherwise, the unpack shader has the wrong idea about how to sample its input (in this case, it thinks that the number of rows in the texture is 10, when it should be 12). This causes inaccurate values to be passed to the subsequent op, which was caught by several `tfjs-layers` tests: https://github.com/tensorflow/tfjs-layers/blob/master/src/layers/convolutional_test.ts#L823

**Comment by annxingyuan** [comment]
Expensive reshapes are ones that require changing the texture layout. In theory we can't say whether 1x4 or 2x2 packing causes fewer expensive reshapes, because some reshapes are expensive for 2x2 and free for 1x4, and vice versa (e.g. `[3, 2, 4] --> [12, 2]` is expensive for 1x4 packing, but free for 2x2 packing). However for the models I looked at 1x4 packing causes fewer expensive reshapes:

### MobileNet v2

total reshapes: *367*
expensive reshapes (2x2): *16*
expensive reshapes (1x4): *0*

### MobileNet v1

total reshapes: *162*
expensive reshapes (2x2): *4*
expensive reshapes (1x4): *0*

### COCO SSD

total reshapes: *1010*
expensive reshapes (2x2): *203*
expensive reshapes (1x4): *126*

### PoseNet

total reshapes: *215*
expensive reshapes (2x2): *34*
expensive reshapes (1x4): *0*

## Notes

For MobileNet v1 and v2, all but one of the expensive reshapes are caused by going back and forth between `matMul` reshaping its input to `2D` and `batchNormalization` reshaping its input to `4D` (`[1, 7, 7, 160] <--> [49, 160]`). For COCO SSD, ~75% of the expensive reshapes are `[1917, 1] <--> [1917]` for the `greater` op. 

cc @nsthorat @dsmilkov
